src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -320,20 +315,16 @@ private void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
             {
                 if (entries[i].hashCode >= 0)
                 {
-                    array[index++] = new KeyValuePair<TKey, TValue>(entries[i].key, entries[i].value);
+                    array[index + i] = new KeyValuePair<TKey, TValue>(entries[i].key, entries[i].value);
```

Comment:
This is a breaking change and needs to be reverted.  Let's say there were 2 entries in the dictionary and then 1 was removed, so for example Count == 1, entries[0].hashCode == -1, entries[1] == the item.  This would previously copy the single entry to array[0]; now it's going to copy it to array[1], which a) is the wrong index and b) will blow up if a developer passed in an array sized to Count.
cc: @jkotas, @benaadams

Reply:
due to the skip :-/

Give a few mins...

Reply:
(My analysis isn't quite right, but something is off here.)

Reply:
`i` always increases but there might not be an entry (due to `hashCode < 0`) so the index for the array shouldn't increase when there isn't an entry.

Reply:
Right

Reply:
> _count isn't the same as Count (Count is confusingly _count - _freeCount). Lots of System.Net.Http outerloop tests are failing as a result.  Looks like we must also then have a Dictionary test hole that needs to be plugged.

Because they should have failed with array out of bounds on CopyTo?

Reply:
> Because they should have failed with array out of bounds on CopyTo?

That's the failure mode hitting System.Net.Http, as CurlHandler uses CopyTo.

But if the output array is large enough to accommodate it, the bug could also manifest as a gap in the copied data.

Reply:
https://github.com/dotnet/coreclr/pull/17300